### PR TITLE
A sample program  : how to use ORB AKAZE BRISK descriptors for matching.

### PR DESCRIPTION
**Pull request by LaurentBerger**
see http://answers.opencv.org/question/61062/features2d-example/

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
Thank you for the example, but could you please add some lines of comments inside the code. For a newly starting user, reading such a block of code is quite difficult. Also it is the case for most samples that if you run the sample without parameters, that it gives a brief description of what the example does. Could you add this?

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    typeAlgoMatch.push_back("BruteForce");
+    typeAlgoMatch.push_back("BruteForce-Hamming");
+    typeAlgoMatch.push_back("BruteForce-Hamming(2)");
+
+    vector<String> typeDesc;
+    typeDesc.push_back("AKAZE");
+    typeDesc.push_back("ORB");
+    typeDesc.push_back("BRISK");
+
+    String  dataFolder("../data/");
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Let users pass this as a parameter, hard coded paths are not desired in samples.

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include <vector>
+#include <iostream>
+
+using namespace std;
+using namespace cv;
+
+
+int main(void)
+{
+    vector<String> typeAlgoMatch;
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Add a description here on what the matching algorithms are about.

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    typeAlgoMatch.push_back("BruteForce");
+    typeAlgoMatch.push_back("BruteForce-Hamming");
+    typeAlgoMatch.push_back("BruteForce-Hamming(2)");
+
+    vector<String> typeDesc;
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Add a description here on what the descriptor algorithms are about.

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    typeDesc.push_back("AKAZE");
+    typeDesc.push_back("ORB");
+    typeDesc.push_back("BRISK");
+
+    String  dataFolder("../data/");
+    vector<String> fileName;
+    fileName.push_back("basketball1.png");
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Might want to leave a notice here that these images can be found in the data folder under the samples

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    fileName.push_back("basketball2.png");
+
+    Mat img1 = imread(dataFolder+fileName[0], IMREAD_GRAYSCALE);
+    Mat img2 = imread(dataFolder+fileName[1], IMREAD_GRAYSCALE);
+
+    Ptr<Feature2D> b;
+
+    vector<String>::iterator itDesc;
+// Descriptor loop
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Indentation is wrong here. Please add some extra spaces

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    for (itDesc = typeDesc.begin(); itDesc != typeDesc.end(); itDesc++){
+        Ptr<DescriptorMatcher> descriptorMatcher;
+        vector<DMatch> matches;	        /*<! Match between img and img2*/
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Normally comments are posted just above a line of code using the `//` command.

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            b = ORB::create();
+        }
+        else if (*itDesc == "BRISK"){
+            b = BRISK::create();
+        }
+        try {
+            b->detect(img1, keyImg1, Mat());
+            b->compute(img1, keyImg1, descImg1);
+            b->detectAndCompute(img2, Mat(),keyImg2, descImg2,false);
+ // Match method loop
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: fix identation

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+             for (itMatcher = typeAlgoMatch.begin(); itMatcher != typeAlgoMatch.end(); itMatcher++){
+                descriptorMatcher = DescriptorMatcher::create(*itMatcher);
+                descriptorMatcher->match(descImg1, descImg2, matches, Mat());
+ // Keep best matches only to have a nice drawing
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: fix identation

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                Mat index;
+                int nbMatch=int(matches.size());
+                Mat tab(nbMatch, 1, CV_32F);
+                for (int i = 0; i<nbMatch; i++)
+                    tab.at<float>(i, 0) = matches[i].distance;
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: It seems useless, but I prefer to add brackets even with a single execution command. For new users it is quite confusing to not find these brackets.

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                sortIdx(tab, index, SORT_EVERY_COLUMN + SORT_ASCENDING);
+                vector<DMatch> bestMatches;
+                for (int i = 0; i<30; i++)
+                    bestMatches.push_back(matches[index.at<int>(i, 0)]);
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Same comment as above

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                drawMatches(img1, keyImg1, img2, keyImg2, bestMatches, result);
+                namedWindow(*itDesc+": "+*itMatcher, WINDOW_AUTOSIZE);
+                imshow(*itDesc + ": " + *itMatcher, result);
+                FileStorage fs(*itDesc+"_"+*itMatcher+"_"+fileName[0]+"_"+fileName[1]+".xml", FileStorage::WRITE);
+                fs<<"Matches"<<matches;
+                vector<DMatch>::iterator it;
+                cout << "Index \tIndex \tindex \tdistance\n";
+                cout << "in img1\tin img2\timage\t\n";
+                for (it = matches.begin(); it != matches.end(); it++)
+                    cout << it->queryIdx << "\t" << it->trainIdx << "\t" << it->imgIdx << "\t" << it->distance<<"\n";
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Same comment as above

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                waitKey();
+            }
+        }
+        catch (Exception& e){
+            cout << "Feature : " << *itDesc << "\n";
+            if (itMatcher != typeAlgoMatch.end())
+                cout << "Matcher : " << *itMatcher << "\n";
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: Same here

**Comment by StevenPuttemans** [comment]
Added some small comments for your knowledge!

**Comment by StevenPuttemans** [comment]
Thank you! Also before a PR gets incorporated, you should squash all your commits into a single one. This can be done with the `git rebase -i HEAD~7` command!

File: samples/cpp/matchmethod_orb_akaze_brisk.cpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+
+
+int main(int argc, char *argv[])
+{
+    vector<String> typeDesc;
+    vector<String> typeAlgoMatch;
+    vector<String> fileName;
+    help();
+    system("cd");
```
Comment (StevenPuttemans) [inline samples/cpp/matchmethod_orb_akaze_brisk.cpp]: change this line to `int ret = system("cd");` to remove the warning about the return value not being used

**Comment by vpisarev** [comment]
@LaurentBerger, I think, the sample is in good shape. Could you squash the commits as @StevenPuttemans suggests? Then we'll merge the PR

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**